src/app_main.rs

```diff
@@ -189,38 +201,47 @@ where
 
     fn render(&mut self) {
         let fragment = self.app.fragment();
-        if self.pgpu_state.is_none() {
-            let handle = &self.handle;
-            let scale = handle.get_scale().unwrap();
-            let insets = handle.content_insets().to_px(scale);
-            let mut size = handle.get_size().to_px(scale);
-            size.width -= insets.x_value();
-            size.height -= insets.y_value();
+        let handle = &self.handle;
+        let scale = handle.get_scale().unwrap_or_default();
+        let insets = handle.content_insets().to_px(scale);
+        let mut size = handle.get_size().to_px(scale);
+        size.width -= insets.x_value();
+        size.height -= insets.y_value();
+        let width = size.width as u32;
+        let height = size.height as u32;
+        if self.surface.is_none() {
             println!("render size: {:?}", size);
-            self.pgpu_state = Some(
-                crate::render::PgpuState::new(
-                    handle,
-                    handle,
-                    size.width as usize,
-                    size.height as usize,
-                )
-                .unwrap(),
-            );
+            self.surface = Some(self.render_cx.create_surface(handle, width, height));
```

Comment:
`create_surface` laundering the unsafety of `RawWindowHandle` isn't ideal.

Really the ecosystem needs to get their act together and find a safe way to access swapchains from wgpu.
I don't really want to block on it, because afaik the safety comment on literally every call to `instance.create_surface` is along the lines of "hope this works :)"

Reply:
Yeah, I suppose the assumption is that if you're sound on both ends of `RawWindowHandle` then you're good. Otherwise, there be dragons. Still, it's a pretty nifty solution to this problem that avoids the need to bless a particular windowing library or marshal these types around yourself.

Reply:
I think the issue is that the current raw-window-handle doesn't guarantee that the window will be sound to create a swapchain for, whilst wgpu assumes it is.
I don't know when this breaks down, but that's the real concern. It's not just about soundness on each side, it's about the windows having the expected capabilities.